File: typescript/problem_21.ts
// minRooms returns the minimum number of rooms to schedule all intervals.
// Sort by start time; use a min-heap of end times.
// This runs in O(n log n) time.
export function minRooms(intervals: [number, number][]): number {
    if (intervals.length === 0) return 0;
    const sorted = [...intervals].sort((a, b) => a[0] - b[0]);
    const ends: number[] = []; // kept sorted as a simple min-heap substitute

    for (const [start, end] of sorted) {
        if (ends.length > 0 && ends[0] <= start) {
            ends.shift(); // reuse earliest-ending room
        }
        // insert end keeping array sorted
        const pos = ends.findIndex(e => e > end);
        if (pos === -1) ends.push(end); else ends.splice(pos, 0, end);
    }
    return ends.length;
}

export function problem21(): void {
    console.log("Problem 21:");
    if (minRooms([[30,75],[0,50],[60,150]]) !== 2) throw new Error("assertion failed");
    if (minRooms([[0,10],[10,20]])          !== 1) throw new Error("assertion failed");
    if (minRooms([[0,10],[0,10]])           !== 2) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: lua/problem_21.lua
-- minRooms returns the minimum number of rooms to schedule all intervals.
-- Sort by start time; maintain a sorted list of end times (min-heap substitute).
-- This runs in O(n log n) time.
local function minRooms(intervals)
    if #intervals == 0 then return 0 end
    table.sort(intervals, function(a, b) return a[1] < b[1] end)
    local ends = {}  -- sorted list of end times

    for _, iv in ipairs(intervals) do
        local start, finish = iv[1], iv[2]
        if #ends > 0 and ends[1] <= start then
            table.remove(ends, 1)  -- reuse earliest-ending room
        end
        -- insert finish keeping ends sorted
        local pos = #ends + 1
        for i = 1, #ends do
            if ends[i] > finish then pos = i; break end
        end
        table.insert(ends, pos, finish)
    end
    return #ends
end

local function problem21()
    print("Problem 21:")
    assert(minRooms({{30,75},{0,50},{60,150}}) == 2, "assertion failed")
    assert(minRooms({{0,10},{10,20}})          == 1, "assertion failed")
    assert(minRooms({{0,10},{0,10}})           == 2, "assertion failed")
    print("All Tests Passed")
end

return { problem21 = problem21 }
